### ISSUE TITLE
sampling from `PSDMat` results in different values as from `PDMat` and empirical statistics doesn't match the analytical results

### ISSUE THREAD

**Issue by wytbella** [invenia/PDMatsExtras.jl#11]
The example below shows that even if the same strictly positive definite matrix is wrapped in `PDMats` and `PSDMats` before wrapping into a matrix, they result in different samples when drawing from the distribution.

In an [internal use case](https://gitlab.invenia.ca/invenia/Metrics.jl/-/merge_requests/103#note_212080), we also found that the emperical results based on `PDMat` converges to analytical solution much better than the one based on `PSDMat`

```julia
julia> using PDMats, PDMatsExtras, LinearAlgebra, Distributions


julia> using Random

julia> using Distributions: GenericMvTDist

julia> seed = 1
1

julia> num_samples = 5_000_000
5000000

julia> B = (reshape(2:10, 3, 3) / 12) .^ 2
3×3 Array{Float64,2}:
 0.0277778  0.173611  0.444444
 0.0625     0.25      0.5625
 0.111111   0.340278  0.694444

julia> A = B' * B + I
3×3 Array{Float64,2}:
 1.01702    0.0582562  0.124662
 0.0582562  1.20843    0.45409
 0.124662   0.45409    1.99619

julia> # construct P(S)DMats
       pd = PDMat(Symmetric(A))
3×3 PDMat{Float64,Array{Float64,2}}:
 1.01702    0.0582562  0.124662
 0.0582562  1.20843    0.45409
 0.124662   0.45409    1.99619

julia> psd = PSDMat(Symmetric(A))
3×3 PSDMat{Float64,Array{Float64,2}}:
 1.01702    0.0582562  0.124662
 0.0582562  1.20843    0.45409
 0.124662   0.45409    1.99619

julia> # psd = PSDMat(Matrix(cholesky(Symmetric(A))))

       # MvNormal
       mvn_pd = MvNormal(ones(size(pd, 1)), pd)
FullNormal(
dim: 3
μ: [1.0, 1.0, 1.0]
Σ: [1.0170235339506173 0.05825617283950618 0.12466242283950618; 0.05825617283950618 1.2084297839506173 0.4540895061728396; 0.12466242283950618 0.4540895061728396 1.996190200617284]
)


julia> mvn_psd = MvNormal(ones(size(psd, 1)), psd)
MvNormal{Float64,PSDMat{Float64,Array{Float64,2}},Array{Float64,1}}(
dim: 3
μ: [1.0, 1.0, 1.0]
Σ: [1.0170235339506173 0.05825617283950618 0.12466242283950618; 0.05825617283950618 1.2084297839506173 0.4540895061728396; 0.12466242283950618 0.4540895061728396 1.996190200617284]
)


julia> mvn_sam_pd = rand(MersenneTwister(seed), mvn_pd, num_samples);

julia> mvn_sam_psd = rand(MersenneTwister(seed), mvn_psd, num_samples);

julia> mvn_sam_pd == mvn_sam_psd
false

julia> # MvT
       mvt_pd = GenericMvTDist(2.2, ones(size(pd, 1)), pd)
GenericMvTDist{Float64,PDMat{Float64,Array{Float64,2}},Array{Float64,1}}(
df: 2.2
dim: 3
μ: [1.0, 1.0, 1.0]
Σ: [1.0170235339506173 0.05825617283950618 0.12466242283950618; 0.05825617283950618 1.2084297839506173 0.4540895061728396; 0.12466242283950618 0.4540895061728396 1.996190200617284]
)


julia> mvt_psd = GenericMvTDist(2.2, ones(size(psd, 1)), psd)
GenericMvTDist{Float64,PSDMat{Float64,Array{Float64,2}},Array{Float64,1}}(
df: 2.2
dim: 3
μ: [1.0, 1.0, 1.0]
Σ: [1.0170235339506173 0.05825617283950618 0.12466242283950618; 0.05825617283950618 1.2084297839506173 0.4540895061728396; 0.12466242283950618 0.4540895061728396 1.996190200617284]
)


julia> mvt_sam_pd = rand(MersenneTwister(seed), mvt_pd, num_samples);

julia> mvt_sam_psd = rand(MersenneTwister(seed), mvt_psd, num_samples);

julia> mvt_sam_pd == mvt_sam_psd
false
```

**Comment by willtebbutt** [comment]
Could you show the numbers @wytbella ? Is it that the sampled values are completely different, or just off by a little bit?

**Comment by AlexRobson** [comment]
I can reproduce this. The numbers do look different.
```
d1 = MvNormal(zeros(3), PDMat(A))
d2 = MvNormal(zeros(3), PSDMat(A))
```
```
julia> rand(MersenneTwister(1), d1, 3)
3×3 Array{Float64,2}:
  0.299808  -0.0105338   2.31454
  0.436955  -0.92166    -2.35615
 -0.612737   0.0762564   0.0746854
julia> rand(MersenneTwister(1), d2, 3)
3×3 Array{Float64,2}:
  0.420028  -0.0147577   3.24265
  0.497542  -0.885387   -1.64565
 -0.563042   0.287636    0.670222
 ```
One thing to note, in `PSDMat` it uses a pivoted cholesky, in PDMats, it uses a standard cholesky, which means we have a different unwhitening transform in the rand call in [Distributions](https://github.com/JuliaStats/Distributions.jl/blob/863844c88e4153af13996f571fcc612d159de542/src/multivariate/mvnormal.jl#L275)

```
julia> PDMats.unwhiten!(d2.Σ, ones(3,))
3-element Array{Float64,1}:
 1.4128659528126806
 1.3726497452380335
 1.1208801406990954

julia> PDMats.unwhiten!(d1.Σ, ones(3,))
3-element Array{Float64,1}:
 1.0084758469842583
 1.155533745084465
 1.8780308620173645
```

So...IIUC, we wouldn't expect the draws to be the same. However, there is no reason for the convergence to not be the same.

**Comment by willtebbutt** [comment]
Perhaps we could check that the mean and cov esimated empirically from samples converge to the same thing?

**Comment by willtebbutt** [comment]
Actually, I wonder whether this is actually a bug? Specifically, shouldn't there be something to account for the pivots [here](https://github.com/invenia/PDMatsExtras.jl/blob/f58218466471ddf00206533fb90227975132341a/src/psd_mat.jl#L83) if we're using the pivoted transform?

**Comment by AlexRobson** [comment]
> Specifically, shouldn't there be something to account for the pivots here if we're using the pivoted transform?

I was looking at that too. If we define something like:

```
function PDMats.unwhiten!(r::StridedVecOrMat, a::PSDMat, x::StridedVecOrMat)
    #ip = invperm(a.chol.)
    cf = a.chol.P' * a.chol.U * a.chol.P
    v = PDMats._rcopy!(r, x)
    mul!(similar(v), cf, v)
    #istriu(cf) ? lmul!(transpose(cf), v) : lmul!(cf, v)
end
```

It improves things a bit (first row is c;pser), but it's not identical, so im wondering if i have something wrong in the set up:

```
julia> rand(MersenneTwister(1), d1, 3)
3×3 Array{Float64,2}:
  0.299808  -0.0105338   2.31454
  0.436955  -0.92166    -2.35615
 -0.612737   0.0762564   0.0746854

julia> rand(MersenneTwister(1), d2, 3)
3×3 Array{Float64,2}:
  0.298538  -0.0104892   2.30474
  0.41045   -0.882327   -2.31801
 -0.695246   0.168977    0.222642
 ```

**Comment by willtebbutt** [comment]
Ah, if you replace
```julia
cf = a.chol.P' * a.chol.U * a.chol.P
```
with
```julia
cf = a.chol.U * a.chol.P'
```
I think things will work out as if you compute `cf'cf` it recovers the original matrix.

**Comment by AlexRobson** [comment]
Closer. 

This is what `master` is doing now i believe (you see something similar with an arbitrary covariance matrix of the form `Σ = r' * r` where `r = rand(10,10)`. 

```
julia> 100 * (cov(rand(MersenneTwister(1), d1, 1000000); dims = 2) .- d1.Σ) ./ d1.Σ
3×3 Array{Float64,2}:
 -0.0478077   0.450099    1.00101
  0.450099   -0.0854778  -0.0857939
  1.00101    -0.0857939  -0.133285

julia> 100 * (cov(rand(MersenneTwister(1), d2, 1000000); dims = 2) .- d1.Σ) ./ d1.Σ

3×3 Array{Float64,2}:
  96.1838   679.766        0.96575
 679.766     -0.0702238  -87.1269
   0.96575  -87.1269     -49.1209
  ```
  
  Using (looking at what LinearAlgebra is doing [here](https://github.com/JuliaLang/julia/blob/master/stdlib/LinearAlgebra/src/cholesky.jl#L432)
  
  ```
  function PDMats.unwhiten!(r::StridedVecOrMat, a::PSDMat, x::StridedVecOrMat)
    ip = invperm(a.chol.p)
    cf = a.chol.U[1:a.chol.rank, ip]
    v = PDMats._rcopy!(r, x)
    mul!(similar(v), transpose(cf), v)
    #istriu(cf) ? lmul!(transpose(cf), v) : lmul!(cf, v)
end
```

This set-up recovers the full matrix with `collect(cf') * cf == d2.

```
julia> 100 * (cov(rand(MersenneTwister(1), d1, 1000000); dims = 2) .- d1.Σ) ./ d1.Σ
3×3 Array{Float64,2}:
 -0.0478077   0.450099    1.00101
  0.450099   -0.0854778  -0.0857939
  1.00101    -0.0857939  -0.133285

julia> 100 * (cov(rand(MersenneTwister(1), d2, 1000000); dims = 2) .- d1.Σ) ./ d1.Σ
3×3 Array{Float64,2}:
 -0.135633   0.341778    0.96575
  0.341778  -0.0702238   0.0378924
  0.96575    0.0378924  -0.0478077
```

But it's still not exact. I'm a bit unsure right now if that's something we can expect.

**Comment by willtebbutt** [comment]
Yeah, I don't think that it is -- they're both valid square roots, but they are different, so individual samples will be different, they ought to be the same in distribution 🤷 I'm actually not sure what the best way to test for convergence is -- probably some kind of simple hypothesis test would suffice, but we'd need to do some maths.

**Comment by olivierverdier** [comment]
Oh, I see, so the problem is indeed ultimately the call to the pivoted Cholesky algorithm. The solution to fix `unwhiten` is in the documentation: `A ≈ Up' * Up ≈ Lp * Lp' with Up = F.U[1:F.rank, invperm(F.p)] and Lp = F.L[invperm(F.p), 1:F.rank]`, so I guess `unwhiten` should essentially use (at the last row):
```julia
res = istriu(cf) ? lmul!(transpose(cf), v) : lmul!(cf, v)
return res[a.chol.p] # but should permute on first dimension only
```

**Comment by oxinabox** [comment]
Closed by #33
